diffkemp/simpll/SMTBlockComparator.cpp

```diff
@@ -0,0 +1,842 @@
+        InstL++;
+    }
+
+    throw NoSynchronizationPointException();
+}
+
+bool SMTBlockComparator::isSupportedInst(BasicBlock::const_iterator Inst) {
+    if (isDebugInfo(*Inst))
+        return true;
+    switch (Inst->getOpcode()) {
+    // Unary operators
+    case Instruction::FNeg:
+    // Binary operators
+    case Instruction::Add:
+    case Instruction::FAdd:
+    case Instruction::Sub:
+    case Instruction::FSub:
+    case Instruction::Mul:
+    case Instruction::FMul:
+    case Instruction::UDiv:
+    case Instruction::SDiv:
+    case Instruction::FDiv:
+    case Instruction::URem:
+    case Instruction::SRem:
+    case Instruction::FRem:
+    // Bitwise binary operators
+    case Instruction::Shl:
+    case Instruction::LShr:
+    case Instruction::AShr:
+    case Instruction::And:
+    case Instruction::Or:
+    case Instruction::Xor:
+    // Casts
+    case Instruction::Trunc:
+    case Instruction::ZExt:
+    case Instruction::SExt:
+    case Instruction::FPTrunc:
+    case Instruction::FPExt:
+    case Instruction::FPToUI:
+    case Instruction::FPToSI:
+    case Instruction::UIToFP:
+    case Instruction::SIToFP:
+    // Other
+    case Instruction::Call:
+    case Instruction::ICmp:
+    case Instruction::FCmp:
+    case Instruction::Select:
+        return true;
+    default:
+        return false;
+    }
+}
+
+z3::expr SMTBlockComparator::createVar(z3::context &c,
+                                       const char *name,
+                                       Type *type) {
+    if (type->isDoubleTy()) {
+        return c.fpa_const(name, 11, 53);
+    } else if (type->isFloatTy()) {
+        return c.fpa_const(name, 8, 24);
+    } else if (type->isIntegerTy()) {
+        if (type->getIntegerBitWidth() == 1) {
+            return c.bool_const(name);
+        } else {
+            return c.bv_const(name, type->getIntegerBitWidth());
+        }
+    }
+    std::string msg = "Unsupported operand type " + typeToString(type);
+    throw UnsupportedOperationException(msg);
+}
+
+z3::expr SMTBlockComparator::createConstant(z3::context &c,
+                                            const Constant *constant) {
+    if (constant->getType()->isIntegerTy()) {
+        auto value = dyn_cast<ConstantInt>(constant)->getSExtValue();
+        auto bitWidth = constant->getType()->getIntegerBitWidth();
+        if (bitWidth == 1) {
+            return c.bool_val(value);
+        } else {
+            return c.bv_val(value, bitWidth);
+        }
+    } else if (constant->getType()->isFloatTy()) {
+#if LLVM_VERSION_MAJOR < 11
+        return c.fpa_val(
+                dyn_cast<ConstantFP>(constant)->getValueAPF().convertToFloat());
+#else
+        return c.fpa_val(
+                dyn_cast<ConstantFP>(constant)->getValue().convertToFloat());
+#endif
+    } else if (constant->getType()->isDoubleTy()) {
+#if LLVM_VERSION_MAJOR < 11
+        return c.fpa_val(
+                dyn_cast<ConstantFP>(constant)->getValueAPF().convertToFloat());
```

Comment:
Shouldn't it be
```suggestion
                dyn_cast<ConstantFP>(constant)->getValueAPF().convertToDouble());
```
?